### PR TITLE
Add support for x.509 Name Serial Number attribute in subject of certificates

### PR DESCRIPTION
**Pull request by mwielgoszewski**
This PR adds support for specifying X.509 Name serialNumber attributes in subjects. It is possible to already to do this using a custom OID (2.5.4.5) with `allowed_other_sans="2.5.4.5;utf8:*" in the role configuration, however this field is not regularly parsed by various applications. It's also possible to issue certificates by passing a CSR to the sign-verbatim endpoint, but as the documents suggest it is a highly trusted and potentially dangerous endpoint. However, the sign-verbatim endpoint does not respect role specific requirements (such as EKU restrictions) and submitted CSR's may override/bypass them.

Discussion here: https://github.com/hashicorp/vault/issues/4562#issuecomment-393907646

### CODE REVIEW

File: builtin/logical/pki/ca_util.go [hunk 1]
```diff
@@ -29,20 +29,21 @@ func (b *backend) getGenerationParams(
...
-		StreetAddress:    data.Get("street_address").([]string),
-		PostalCode:       data.Get("postal_code").([]string),
+		TTL:                  time.Duration(data.Get("ttl").(int)) * time.Second,
+		KeyType:              data.Get("key_type").(string),
+		KeyBits:              data.Get("key_bits").(int),
+		AllowLocalhost:       true,
+		AllowAnyName:         true,
+		AllowIPSANs:          true,
+		EnforceHostnames:     false,
+		AllowedSerialNumbers: []string{"*"},
```
Comment (jefferai) [inline builtin/logical/pki/ca_util.go]: This should not be set to allow any by default. The user should need to explicitly add that. (There are two other places this needs to be changed). Also it doesn't actually read `allowed_serial_numbers` :-/

Comment (jefferai) [reply builtin/logical/pki/ca_util.go]: Ah, never mind. I thought this was in path_roles, I skipped around too much.

File: builtin/logical/pki/cert_util.go [hunk 2]
```diff
@@ -695,18 +718,26 @@ func generateCreationBundle(b *backend, data *dataBundle) error {
...
+			ridSerialNumber = data.csr.Subject.SerialNumber
 		}
 		if cn == "" {
 			cn = data.apiData.Get("common_name").(string)
 			if cn == "" && data.role.RequireCN {
 				return errutil.UserError{Err: `the common_name field is required, or must be provided in a CSR with "use_csr_common_name" set to true, unless "require_cn" is set to false`}
 			}
 		}
+		if ridSerialNumber == "" {
+			ridSerialNumberRaw, ok := data.apiData.GetOk("serial_number")
```
[the remark this answers is not in the document]
Comment (mwielgoszewski) [reply builtin/logical/pki/cert_util.go]: This was failing otherwise since it's not a field common to CA generation/signing.

**Comment by jefferai** [comment]
Looking good, just two minor comments.

File: builtin/logical/pki/cert_util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jefferai) [reply builtin/logical/pki/cert_util.go]: Feels like it should be? I don't see a reason to not allow specifying the serial number when generating a CA.

File: builtin/logical/pki/backend_test.go [hunk 3]
```diff
@@ -3221,6 +3221,129 @@ func TestBackend_OID_SANs(t *testing.T) {
 	t.Logf("certificate 3 to check:\n%s", certStr)
 }
 
+func TestBackend_AllowedSerialNumbers(t *testing.T) {
+	coreConfig := &vault.CoreConfig{
+		LogicalBackends: map[string]logical.Factory{
+			"pki": Factory,
+		},
+	}
+	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
```
Comment (jefferai) [inline builtin/logical/pki/backend_test.go]: You're one of the first outside contributors to actually use NewTestCluster. Kudos!

File: builtin/logical/pki/cert_util.go [hunk 4]
```diff
@@ -708,6 +732,13 @@ func generateCreationBundle(b *backend, data *dataBundle) error {
 			}
 		}
 
+		if data.csr != nil && len(data.role.AllowedSerialNumbers) > 0 {
+			ridSerialNumber = data.csr.Subject.SerialNumber
+		}
+		if ridSerialNumber == "" {
```
Comment (jefferai) [inline builtin/logical/pki/cert_util.go]: I like the check that you added but think the order should be reversed -- API-specified values should override CSR values. Part of the reason is that a lot of people are not careful when constructing CSRs and may pick the same value over and over (e.g. just set something in a template) or may not use a sufficiently random source if they aren't intending for it to be something specific.

File: builtin/logical/pki/backend_test.go (hunk 3 above)
Comment (mwielgoszewski) [reply builtin/logical/pki/backend_test.go]: I gotta admit, most of this was copy/paste from existing tests 😳

**Comment by jefferai** [comment]
We pushed back 0.10.2 slightly so should be able to get this in for it! Just need that one check to be reversed and I think it's 🌈

File: builtin/logical/pki/backend_test.go (hunk 3 above)
Comment (jefferai) [reply builtin/logical/pki/backend_test.go]: Yes, but, many people copy/paste from existing tests using framework.TestCase and friends, and...we don't like that. (It was a good idea at the time when Vault was very, very simple, and has not at all scaled.) So it's sad when someone comes with a whole ton of TestCase code and we have to decide whether to tell them to redo it all or continue feeding the beast.

**Comment by jefferai** [comment]
Can I just say that I hate that pkix subjects and certs themselves both have serial numbers.

**Comment by jefferai** [comment]
Thank you!

**Pull request merged**